# Predictive Analysis

In [1]:
import sys
import os
import warnings
warnings.filterwarnings('ignore')

In [2]:
import IPython
# from IPython.display import Image
# from IPython.core.display import HTML
from IPython.display import clear_output
from IPython.display import display
from tqdm.notebook import tqdm # FOR FANCY GREEN BAR

In [3]:
import numpy as np
import pandas as pd
import polars as pr # new pkg similar to pandas but faster
import glob

In [43]:
import matplotlib.pyplot as plt
import matplotlib as mpl
from matplotlib import cm
import seaborn as sns
import plotly

# plt.rcParams['figure.figsize'] = [12, 8]
plt.rcParams['figure.dpi'] = 100 # e.g. default 100 but 300 would be a really fine plot, but slower
plt.style.use('fivethirtyeight')

custom = {"axes.edgecolor": "black", "grid.linestyle": "dashed", "grid.color": "red"}
sns.set_style("white", rc = custom)

In [5]:
import re
import random
import time
from datetime import datetime as dt
import scipy
import statsmodels.api as sm

In [6]:
import sklearn
from sklearn import datasets
from sklearn.utils import shuffle
from sklearn.preprocessing import OneHotEncoder, LabelEncoder, StandardScaler

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from xgboost import XGBRegressor

from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, jaccard_score, f1_score, r2_score, roc_curve, auc, log_loss, classification_report

classifiers = {
    "LogisticRegression" : LogisticRegression(),
    "KNeighbors" : KNeighborsClassifier(),
    "SVC" : SVC(),
    "DecisionTree" : DecisionTreeClassifier(),
    "RandomForest" : RandomForestClassifier(),
    "XGBoost" : XGBClassifier()
}

In [7]:
from lazypredict.Supervised import LazyClassifier, LazyRegressor

# Data

In [8]:
random.seed(42)

In [9]:
# True labels
y_true = [0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0]  

# Predicted labels
y_pred= [1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1] 


In [10]:
# Confusion Matrix
cm = confusion_matrix(y_true, y_pred)
# Accuracy
accuracy = accuracy_score(y_true, y_pred)
# Precision
precision = precision_score(y_true, y_pred)
# Recall
recall = recall_score(y_true, y_pred)
# F1-Score
f1 = f1_score(y_true, y_pred)
# ROC Curve and AUC
fpr, tpr, thresholds = roc_curve(y_true, y_pred)
roc_auc = auc(fpr, tpr)

# Jaccurred
jc = jaccard_score(y_true, y_pred)


# print(Confusion Matrix)
# print(cm)
print(f" Accuracy \t{accuracy}")
print(f" Precision \t {precision}")
print(f" Recall \t {recall}")
print(f" F1 \t {f1}")
print(f" ROC \t {roc_auc}")
print(f" Jaccurred \t {jc}")
# print(f"{classification_report(y_test, y_pred)}")

 Accuracy 	0.72
 Precision 	 0.7307692307692307
 Recall 	 0.7307692307692307
 F1 	 0.7307692307692307
 ROC 	 0.7195512820512819
 Jaccurred 	 0.5757575757575758


### Multilabeled

In [11]:
y_true = ['a','b','c','d']
y_pred = ['c','a','c','d']

In [12]:
# Accuracy
accuracy = accuracy_score(y_true, y_pred)
# Precision
precision = precision_score(y_true, y_pred, average="weighted")
# Recal
recall = recall_score(y_true, y_pred, average="weighted")
# F1-Score
f1 = f1_score(y_true, y_pred, average="weighted")
# Jaccurred
jc = jaccard_score(y_true, y_pred, average="weighted")

# ROC Curve and AUC
# fpr, tpr, thresholds = roc_curve(y_true, y_pred)
# roc_auc = auc(fpr, tpr)


print(f" Accuracy \t{accuracy}")
print(f" Precision \t {precision}")
print(f" Recall \t {recall}")
print(f" F1 \t {f1}")
# print(f" ROC \t {roc_auc}")
print(f" Jaccurred \t {jc}")

 Accuracy 	0.5
 Precision 	 0.375
 Recall 	 0.5
 F1 	 0.41666666666666663
 Jaccurred 	 0.375


In [13]:
# # Plot ROC curve
# plt.figure()

# # Plot the ROC curve with a label displaying the ROC AUC score
# plt.plot(fpr, tpr, color='darkorange', lw=2,
#          label='ROC curve (area = %0.2f)' % roc_auc)

# # Plot a dashed diagonal line for reference
# plt.plot([0, 1], [0, 1], color='navy', lw=2, linestyle='--')

# # Set the x and y-axis limits
# plt.xlim([0.0, 1.0])
# plt.ylim([0.0, 1.05])

# # Label the x and y-axes
# plt.xlabel('False Positive Rate')
# plt.ylabel('True Positive Rate')

# # Set the title of the plot
# plt.title('Receiver Operating Characteristic')

# # Add a legend to the plot
# plt.legend(loc='lower right')

# # Display the ROC curve plot
# plt.show()

## Titanic - Machine Learning from Disaster
- https://www.kaggle.com/competitions/titanic/data

In [14]:
print(os.listdir("../notebooks/input_data"))

['bird.png', 'Ekush-Regular.ttf', 'gender_submission.csv', 'test.csv', 'train.csv', 'Umeå1.png', 'Umeå2.png']


In [15]:
df_train = pd.read_csv('../notebooks/input_data/train.csv')
df_test = pd.read_csv('../notebooks/input_data/test.csv')

In [16]:
df_train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male 22.00      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female 38.00      1   
2                               Heikkinen, Miss. Laina  female 26.00      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female 35.00      1   
4                             Allen, Mr. William Henry    male 35.00      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male 27.00      0   
887                       Graham, Miss. Margaret Edith  female 19.00      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male 26.00      0   
890                                Dooley, Mr. Patrick    male 32.00      0   

     Parch            Ticket  Fare Cabin Embarked  
0        0         A/5 21171  7.25   NaN        S  
1        0          PC 17599 71.28   C85        C  
2        0  STON/O2. 3101282  7.92   NaN        S  
3        0            113803 53.10  C123        S  
4        0            373450  8.05   NaN        S  
..     ...               ...   ...   ...      ...  
886      0            211536 13.00   NaN        S  
887      0            112053 30.00   B42        S  
888      2        W./C. 6607 23.45   NaN        S  
889      0            111369 30.00  C148        C  
890      0            370376  7.75   NaN        Q  

[891 rows x 12 columns]

In [17]:
df_test

PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket   Fare Cabin Embarked  
0      male 34.50      0      0              330911   7.83   NaN        Q  
1    female 47.00      1      0              363272   7.00   NaN        S  
2      male 62.00      0      0              240276   9.69   NaN        Q  
3      male 27.00      0      0              315154   8.66   NaN        S  
4    female 22.00      1      1             3101298  12.29   NaN        S  
..      ...   ...    ...    ...                 ...    ...   ...      ...  
413    male   NaN      0      0           A.5. 3236   8.05   NaN        S  
414  female 39.00      0      0            PC 17758 108.90  C105        C  
415    male 38.50      0      0  SOTON/O.Q. 3101262   7.25   NaN        S  
416    male   NaN      0      0              359309   8.05   NaN        S  
417    male   NaN      1      1                2668  22.36   NaN        C  

[418 rows x 11 columns]

In [18]:
df_train.shape, df_test.shape

((891, 12), (418, 11))

- Training set has 891 rows with 12 features and
- Test set has 418 rows with set have 11 features
- One extra feature in training set is **Survived** (target variable)

In [19]:
def concat_df(train_data, test_data):
    # Returns a concatenated df of training and test set
    return pd.concat([train_data, test_data], sort=True).reset_index(drop=True)

def divide_df(all_data):
    # Returns divided dfs of training and test set
    return all_data.loc[:890], all_data.loc[891:].drop(['Survived'], axis=1)

In [45]:
df_all = concat_df(df_train, df_test)

df_train.name = 'Training Set'
df_test.name = 'Test Set'
df_all.name = 'All Set' 

dfs = [df_train, df_test]

print('Number of Training Examples = {}'.format(df_train.shape[0]))
print('Number of Test Examples = {}\n'.format(df_test.shape[0]))
print('Training X Shape = {}'.format(df_train.shape))
print('Training y Shape = {}\n'.format(df_train['Survived'].shape[0]))
print('Test X Shape = {}'.format(df_test.shape))
print('Test y Shape = {}\n'.format(df_test.shape[0]))
print(df_train.columns)
print(df_test.columns)

df_all.shape

Number of Training Examples = 891
Number of Test Examples = 418

Training X Shape = (891, 12)
Training y Shape = 891

Test X Shape = (418, 11)
Test y Shape = 418

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')
Index(['PassengerId', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')


(1309, 12)

## 1. Exploratory Data Analysis¶
### 1.1 Overview
- **PassengerId** is the unique id of the row and it doesn't have any effect on target
- **Survived** is the target variable we are trying to predict (0 or 1):
    - 1 = Survived
    - 0 = Not Survived
- **Pclass (Passenger Class)** is the socio-economic status of the passenger and it is a categorical ordinal feature which has 3 unique values (1, 2 or 3):
    - 1 = Upper Class
    - 2 = Middle Class
    - 3 = Lower Class
- Name, Sex and Age are self-explanatory
- SibSp is the total number of the passengers' siblings and spouse
- Parch is the total number of the passengers' parents and children
- Ticket is the ticket number of the passenger
- Fare is the passenger fare
- Cabin is the cabin number of the passenger
- Embarked is port of embarkation and it is a categorical feature which has 3 unique values (C, Q or S):
    - C = Cherbourg
    - Q = Queenstown
    - S = Southampton

In [26]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [32]:
df_train.describe() # checking numeric indipendent variables

PassengerId  Survived  Pclass    Age  SibSp  Parch   Fare
count       891.00    891.00  891.00 714.00 891.00 891.00 891.00
mean        446.00      0.38    2.31  29.70   0.52   0.38  32.20
std         257.35      0.49    0.84  14.53   1.10   0.81  49.69
min           1.00      0.00    1.00   0.42   0.00   0.00   0.00
25%         223.50      0.00    2.00  20.12   0.00   0.00   7.91
50%         446.00      0.00    3.00  28.00   0.00   0.00  14.45
75%         668.50      1.00    3.00  38.00   1.00   0.00  31.00
max         891.00      1.00    3.00  80.00   8.00   6.00 512.33

## Missing Values

In [33]:
df_train.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [39]:
df_test.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

In [38]:
def display_missing(df):    
    for col in df.columns.tolist():          
        print('{} has missing values \t :: {}'.format(col, df[col].isnull().sum()))
    print('\n')
    
for df in dfs:
    print('{}'.format(df.name))
    display_missing(df)

Training Set
PassengerId has missing values 	 :: 0
Survived has missing values 	 :: 0
Pclass has missing values 	 :: 0
Name has missing values 	 :: 0
Sex has missing values 	 :: 0
Age has missing values 	 :: 177
SibSp has missing values 	 :: 0
Parch has missing values 	 :: 0
Ticket has missing values 	 :: 0
Fare has missing values 	 :: 0
Cabin has missing values 	 :: 687
Embarked has missing values 	 :: 2


Test Set
PassengerId has missing values 	 :: 0
Pclass has missing values 	 :: 0
Name has missing values 	 :: 0
Sex has missing values 	 :: 0
Age has missing values 	 :: 86
SibSp has missing values 	 :: 0
Parch has missing values 	 :: 0
Ticket has missing values 	 :: 0
Fare has missing values 	 :: 1
Cabin has missing values 	 :: 327
Embarked has missing values 	 :: 0




### Obervation

**Age** and **Cabin** has high number of missing values

Missing values in Age are filled with median age, but using median age of the whole data set is not a good choice. Median age of Pclass groups is the best choice because of its high correlation with Age (0.408106) and Survived (0.338481). It is also more logical to group ages by passenger classes instead of other features.

# Reference

- https://scikit-learn.org/stable/api/sklearn.metrics.html
- https://python-charts.com/seaborn/themes/
- https://www.kaggle.com/code/gunesevitan/titanic-advanced-feature-engineering-tutorial
- https://www.kaggle.com/code/alisarpsunay/titanic-survival-prediction-logistic-regression
- https://www.kaggle.com/code/gusthema/titanic-competition-w-tensorflow-decision-forests